MapboxNavigation/NavigationMapView.swift

```diff
@@ -785,6 +756,71 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     
     // MARK: Utility Methods
     
+    /** Modifies the gesture recognizers to also disable course tracking. */
+    func makeGestureRecognizersRespectCourseTracking() {
+        for gestureRecognizer in gestureRecognizers ?? []
+            where gestureRecognizer is UIPanGestureRecognizer || gestureRecognizer is UIRotationGestureRecognizer {
+                gestureRecognizer.addTarget(self, action: #selector(disableUserCourseTracking))
+        }
+    }
+    
+    func makeGestureRecognizersUpdateCourseView() {
+        for gestureRecognizer in gestureRecognizers ?? [] {
+            gestureRecognizer.addTarget(self, action: #selector(updateCourseView(_:)))
+        }
+    }
+    
+    //TODO: Change to point-based distance calculation
+    private func waypoints(on routes: [Route], closeTo point: CGPoint) -> [Waypoint]? {
+        let tapCoordinate = convert(point, toCoordinateFrom: self)
+        let multipointRoutes = routes.filter { $0.routeOptions.waypoints.count >= 3}
+        guard multipointRoutes.count > 0 else { return nil }
+        let waypoints = multipointRoutes.flatMap({$0.routeOptions.waypoints})
+        
+        //lets sort the array in order of closest to tap
+        let closest = waypoints.sorted { (left, right) -> Bool in
+            let leftDistance = left.coordinate.distance(to: tapCoordinate)
+            let rightDistance = right.coordinate.distance(to: tapCoordinate)
+            return leftDistance < rightDistance
+        }
+        
+        //lets filter to see which ones are under threshold
+        let candidates = closest.filter({
+            let coordinatePoint = self.convert($0.coordinate, toPointTo: self)
+            return coordinatePoint.distance(to: point) < tapGestureDistanceThreshold
+        })
+        
+        return candidates
+    }
+    
+    private func routes(closeTo point: CGPoint) -> [Route]? {
+        let tapCoordinate = convert(point, toCoordinateFrom: self)
+        
+        //do we have routes? If so, filter routes with at least 2 coordinates.
+        guard let routes = routes?.filter({ $0.coordinates?.count ?? 0 > 1 }) else { return nil }
+        
+        //Sort routes by closest distance to tap gesture.
+        let closest = routes.sorted { (left, right) -> Bool in
+            
+            //existance has been assured through use of filter.
+            let leftLine = Polyline(left.coordinates!)
+            let rightLine = Polyline(right.coordinates!)
+            let leftDistance = leftLine.closestCoordinate(to: tapCoordinate)!.distance
+            let rightDistance = rightLine.closestCoordinate(to: tapCoordinate)!.distance
```

Comment:
lines `805-809` force unwrap. Yikes! 💀

Reply:
So it's actually not _always_ an anti pattern to use force-unwraps, it is considered good style to use them when you can _guarantee_ an existence check has already been made for the object in question.

In Line 800, we're filtering to ensure that `routes.coordinates != nil`, and that `routes.coordinates.count >= 2`. When we use the bangs in Lines 806 & 807, that is appropriate because we know that `Route.coordinates` exists on those two calls. 

On lines 806-807, we use bangs there because `Polyline.closestCoordinate(to:)` will [return nil](https://github.com/mapbox/turf-swift/blob/master/Sources/Turf/Turf.swift#L260) if the coordinates array passed in is empty. Because we can guarantee that there are at least two coordinates in the polyline, we can assume that the function will never return `nil`.

Reply:
In fact, I noticed that a few minutes after I scrolled back up to review the variable but you had left a comment. So I take it back 😃